### ISSUE TITLE
Static-shape multi-turn support with append-prefill

### ISSUE THREAD

**Issue by mhs4670go** [Samsung/TICO#812]
## Proposal

This is a follow-up to the current static-shape prefill/decode design described in:

https://github.com/Samsung/TICO/issues/586#issuecomment-4161647429

The target NPU only supports static shapes. Also, handling `attention_mask` and `position_embeddings` inside the NPU graph is inefficient, so these are intentionally prepared by the runtime.

Therefore, the multi-turn design should keep the same principle as the existing static prefill/decode design:

- exported decoder-layer graphs use fully static tensor shapes
- KV cache is managed externally by the runtime
- `attention_mask` is prepared by the runtime
- `position_embeddings = (cos, sin)` are prepared by the runtime
- the exported graph does not compute dynamic positions
- the exported graph does not build causal masks
- the exported graph does not own or update the full KV cache

The missing part is efficient multi-turn cache reuse.

Today, after the first prefill and assistant decode, a new user turn can be handled in two simple ways:

1. Re-run prefill on the full conversation history.
2. Feed the new user turn token-by-token through the decode path.

Both are functionally possible, but neither is ideal:

- full-history prefill discards the benefit of the existing KV cache
- decode-loop append requires one graph invocation per user token

For multi-turn serving, we should add a third path:

```text
append-prefill / chunk-prefill
```

This path consumes multiple new user-turn tokens at once and appends their KV tensors to the existing external cache.

---

## High-level runtime flow

The intended multi-turn flow is:

```text
1. prefill(initial prompt)
   - no past KV
   - returns prompt KV
   - runtime writes prompt KV into external cache

2. decode assistant tokens
   - single-token decode
   - runtime updates external cache one token at a time

3. append-prefill(next user turn)
   - consumes multiple new user tokens
   - uses existing external cache as past KV
   - returns only delta KV for the new user tokens
   - runtime appends the delta KV into the external cache

4. decode next assistant tokens
   - continues from the updated cache
```

So the static graph set becomes:

```text
prefill:
  first prompt / full-context recompute path

append-prefill:
  multi-token cache append path for new user turns

decode:
  single-token generation path
```

`append-prefill` should be a separate export mode instead of overloading the existing prefill or decode contract.

---

## Terminology

Let:

```text
B       = batch size
D       = hidden size
Hkv     = number of KV heads
Hd      = head dimension
K       = fixed attention key length, usually max_seq
P       = number of valid tokens already stored in external KV cache
T       = actual number of new user-turn tokens
Q       = static append-prefill block size / bucket size
```

For static-shape export:

```text
T <= Q
P + Q <= K
```

`T` and `P` are runtime metadata. They are not dynamic NPU graph dimensions.

The exported append-prefill graph is specialized for a fixed `Q`.

For example, we may export several bucketed variants:

```text
append_prefill_Q8
append_prefill_Q16
append_prefill_Q32
append_prefill_Q64
```

The existing decode path already covers the `Q = 1` case, so `append_prefill_Q1` is optional.

The runtime selects the smallest usable bucket satisfying:

```text
T <= Q
P + Q <= K
```

If a new user turn is longer than one bucket, the runtime splits it into multiple append-prefill chunks.

---

## Proposed append-prefill decoder-layer contract

For each decoder layer, append-prefill should use the following static-shape contract.

### Inputs

```text
hidden_states:
  shape = (B, Q, D)

attention_mask:
  shape = (B, Q, K)
  additive mask, already prepared by the runtime

position_embeddings:
  cos shape = (B, Q, Hd)
  sin shape = (B, Q, Hd)
  already sliced/gathered by the runtime for the absolute positions of this block

past_key_value:
  past_k shape = (B, Hkv, K - Q, Hd)
  past_v shape = (B, Hkv, K - Q, Hd)
```

### Outputs

```text
hidden_states:
  shape = (B, Q, D)

new_k:
  shape = (B, Hkv, Q, Hd)

new_v:
  shape = (B, Hkv, Q, Hd)
```

The graph returns delta KV only.

It must not return the full present cache.

The runtime writes only the first `T` entries of `new_k/new_v` into the external cache. Padded entries are ignored.

---

## Runtime-managed cache layout

The runtime owns the external KV cache.

A simple compact layout is:

```text
external_k:
  shape = (B, Hkv, K, Hd)

external_v:
  shape = (B, Hkv, K, Hd)
```

Valid cached tokens are packed into the first `P` slots:

```text
external_k[:, :, 0:P, :]
external_v[:, :, 0:P, :]
```

Unused slots are ignored and must remain masked.

For append-prefill with block size `Q`, the graph receives a static past buffer of length `K - Q`:

```text
past_k_input = external_k[:, :, 0:K-Q, :]
past_v_input = external_v[:, :, 0:K-Q, :]
```

The runtime must ensure:

```text
P <= K - Q
```

Inside the graph, the newly computed K/V for the `Q` query tokens are conceptually appended after the past input.

So the attention key layout inside the graph is:

```text
[ past input of length K-Q ][ new block of length Q ]
```

Expanded:

```text
0                         P          K-Q                    K
| valid cached tokens      | masked   | new block tail        |
|--------------------------|----------|-----------------------|
```

The new block is located at the tail positions:

```text
tail_start = K - Q
tail slots = [K-Q, K)
```

After the graph returns `new_k/new_v`, the runtime writes the valid part compactly back into the external cache:

```text
external_k[:, :, P : P + T, :] = new_k[:, :, 0:T, :]
external_v[:, :, P : P + T, :] = new_v[:, :, 0:T, :]
```

Then:

```text
P = P + T
```

The cache slot index itself is not the semantic token position. The semantic position is already encoded by RoPE before the K tensor is stored.

This is the same idea as the existing decode path, where the current token is appended inside the attention computation and then the runtime writes the returned delta KV into the compact external cache.

---

## Runtime-generated position embeddings

The exported graph should not take `position_ids` or compute RoPE positions internally.

Instead, the runtime prepares:

```text
position_embeddings = (cos, sin)
```

with static shapes:

```text
cos: (B, Q, Hd)
sin: (B, Q, Hd)
```

For valid new user tokens:

```text
token i in the append block uses absolute position P + i
```

So for `i < T`:

```text
cos[:, i, :] = rope_cos[P + i]
sin[:, i, :] = rope_sin[P + i]
```

For padded positions `i >= T`, the values are ignored. They only need to be valid tensors with the correct static shape.

No `arange`, `cumsum`, `index_select`, or RoPE table slicing should happen inside the NPU graph.

---

## Runtime-generated attention mask

The exported graph should not build a causal mask internally.

The runtime provides a fully materialized additive attention mask:

```text
attention_mask: (B, Q, K)
```

For each valid query token `i` where `0 <= i < T`:

1. It can attend to all valid previous cached tokens:

```text
[0, P)
```

2. It cannot attend to unused past-cache slots:

```text
[P, K-Q)
```

3. It can attend causally within the new block tail:

```text
[K-Q, K-Q+i]
```

4. It cannot attend to future new-block tokens:

```text
(K-Q+i, K)
```

So the valid key positions for query `i` are:

```text
[0, P) union [K-Q, K-Q+i]
```

All other key positions are filled with the attention mask fill value.

For padded query rows `i >= T`, all positions can remain masked because the runtime ignores the corresponding outputs and delta KV.

Conceptually:

```text
attention_mask[b, i, key] = 0.0         if key is visible
attention_mask[b, i, key] = mask_value  otherwise
```

This keeps the graph static while allowing the runtime to express:

- current cache length `P`
- actual user turn length `T`
- causal visibility inside the appended block
- padded block entries

---

## Bucketed static append-prefill

Because the NPU requires static shapes, append-prefill should be exported per bucket size.

Example bucket set:

```text
Q in {8, 16, 32, 64}
```

Runtime behavior:

```text
remaining_user_tokens = T_total

while remaining_user_tokens > 0:
    choose Q such that:
      chunk_len <= Q
      P + Q <= K

    prepare hidden_states with shape (B, Q, D)
    prepare position_embeddings with shape (B, Q, Hd)
    prepare attention_mask with shape (B, Q, K)
    prepare past_key_value with shape (B, Hkv, K-Q, Hd)

    run append_prefill_Q

    write new_k/new_v[:, :, 0:chunk_len, :] into external cache
    P += chunk_len
```

If no bucket satisfies `P + Q <= K`, the runtime must apply a context-length policy before continuing.

Possible policies are out of scope for the first version, but examples include:

- stop generation / return an error
- full-context recompute with truncation
- sliding-window cache compaction, if supported later
